tests/zkevm/test_worst_compute.py

```diff
@@ -519,3 +521,91 @@ def test_worst_binop_simple(
         post={},
         blocks=[Block(txs=[tx])],
     )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize("shift_right", [Op.SHR, Op.SAR])
+def test_worst_shifts(
+    blockchain_test: BlockchainTestFiller,
+    pre: Alloc,
+    shift_right: Op,
+):
+    """
+    Test running a block with as many shift instructions with non-trivial arguments.
+    This test generates left-right pairs of shifts to avoid zeroing the argument.
+    The shift amounts are randomly pre-selected from the constant pool of 15 values on the stack.
+    """
+
+    def to_signed(x):
+        return x if x < 2**255 else x - 2**256
+
+    def to_unsigned(x):
+        return x if x >= 0 else x + 2**256
+
+    def shr(x, s):
+        return x >> s
+
+    def shl(x, s):
+        return x << s
+
+    def sar(x, s):
+        return to_unsigned(to_signed(x) >> s)
+
+    def select_shift_amount(shift_fn, v):
+        """Select a shift amount that will produce a non-zero result."""
+        while True:
+            index = rng.randint(0, len(shift_amounts) - 1)
+            sh = shift_amounts[index]
+            new_v = shift_fn(v, sh) % 2**256
+            if new_v != 0:
+                return new_v, index
+
+    def make_dup(i):
+        """Create a DUP instruction to get the i-th shift amount constant from the stack."""
+        return Opcode(0x80 + (len(shift_amounts) - i))
```

Comment:
Is there a nicer way of creating DUP opcodes based on index?

Reply:
Good question, not that I know.

Reply:
That doesn't work or I don't understand it. I need two things here: get the opcode number of the `DUP2` opcode (e.g. `int(Op.DUP2)` or `Op.DUP2.opcode`) and then to construct another DUP having the `int` index number, e.g. having `index=4` I need to "create" `DUP4` instruction.

Reply:
I added a TODO for this and I will propose something later.